### DISCUSSION TITLE
How to get "quarkus.profile" value using Kotlin on Quarkus startup?

### DISCUSSION THREAD

**Discussion by jehrenzweig-leagueapps** [quarkusio/quarkus#26823] [Q&A] [answered]
I'm trying to log the value of the `quarkus.profile` configuration property when Quarkus starts.  Here's my `ApplicationStartup.kt` file:
```
import io.quarkus.runtime.Quarkus
import io.quarkus.runtime.QuarkusApplication
import io.quarkus.runtime.annotations.QuarkusMain
import org.slf4j.LoggerFactory

@QuarkusMain
class ApplicationStartup {
	companion object {
		private val logger = LoggerFactory.getLogger(ApplicationStartup::class.java)

		@JvmStatic
		fun main(args: Array<String>) {
			logger.info("[ApplicationStartup.main()] Entered method call.")
			Quarkus.run(ApplicationInitializer::class.java, *args)
			logger.info("[ApplicationStartup.main()] Exiting method call.")
		}
	}

	class ApplicationInitializer : QuarkusApplication {
		override fun run(vararg args: String): Int {
			logger.info("[ApplicationInitializer.run()] Entered method call.")
			Quarkus.waitForExit()
			logger.info("[ApplicationInitializer.run()] Exiting method call.")
			return 0
		}
	}
}
```

And here's the rudimentary `QuarkusConfiguration.kt` file I've created:
```
import io.smallrye.config.ConfigMapping

@ConfigMapping(prefix = "quarkus")
interface QuarkusConfiguration {
	fun profile(): String
}
```

What changes would I need to make in `ApplicationStartup.kt` that would let me do something like this?
```
logger.info("The value of [quarkus.profile] is {}", quarkusProfileValue)
```

[a comment by an automated account is not shown]

**Comment by jehrenzweig-leagueapps** [comment]
So I think I found a way to do it, by adding the following code at the start of `ApplicationInitializer.run()`:
```
class ApplicationInitializer : QuarkusApplication {
	override fun run(vararg args: String): Int {
		try {
			val quarkusProfileValue = ConfigProvider.getConfig().getValue("quarkus.profile", String::class.java)
			logger.info("[ApplicationInitializer.run()] quarkus.profile = \"{}\".", quarkusProfileValue ?: "<null>")
		} catch (e: Exception) {
			logger.info("[ApplicationInitializer.run()] quarkus.profile = <ERROR>")
		}

		logger.info("[ApplicationInitializer.run()] Entered method call.")
		Quarkus.waitForExit()
		logger.info("[ApplicationInitializer.run()] Exiting method call.")
		return 0
	}
}
```

Which leads me to my next question: is there a better/cleaner way to do this?

**Reply by jehrenzweig-leagueapps** [reply]
So I think I was doing it wrong.  Swapping out this line:
```
val quarkusProfileValue = ConfigProvider.getConfig().getValue("quarkus.profile", String::class.java)
```

For this line:
```
val quarkusProfileValue = io.quarkus.runtime.configuration.ProfileManager.getActiveProfile()
```

Works perfectly now.

**Reply by radcortez** [reply]
The better way to do it is to call `SmallRyeConfig#getProfiles` :)

You either inject `SmallRyeConfig` to get the instance or you need to do `ConfigProvider.getConfig().unwrap(SmallRyeConfig.class)`.

**Reply by jehrenzweig-leagueapps** [reply]
Why is that way better than using `ProfileManager#getActiveProfile`?  I have to either inject an instance of `SmallRyeConfig` -- which I'm not sure how to do in this class in Kotlin -- or unwrap it; all I need to do is import `ProfileManager`.  And if I do use the `SmallRyeConfig#getProfiles` approach, I have to worry about how many items are the array it returns; `ProfileManager` gives me the single active profile, no array tinkering required.

**Reply by radcortez** [reply]
`ProfileManager` is an internal Quarkus API, which may be removed or rewritten at some point. We have been moving most of the config stuff to SmallRye Config. 

For instance, SmallRye Config does support multiple profile activation (that is why you have the List in the API for `getProfiles`), and `ProfileManager` does not. Also, if you set a parent profile with `quarkus.config.profile.parent`, `ProfileManager` does not pick that up.

**Reply by jehrenzweig-leagueapps** [reply]
Interesting... I'd been under the impression that only a single Quarkus profile could be active at any given moment.  Most of the documentation I've seen to date indicates the steps needed to activate a single profile by name, either by setting the `quarkus.profile` system property or the `QUARKUS_PROFILE` environmental variable.

How do I use the info returned by `SmallRyeConfig#getProfiles` to identify the name of the Quarkus profile that's active at runtime when Quarkus starts up?

**Reply by radcortez** [reply, marked as answer]
Yes, Quarkus does not fully support multiple profiles at the moment. The plan is to add them at some point, which is why I do not recommend using `ProfileManager` because it will most likely have to go away (or be changed in an incompatible way). In practice, everything that internally uses `ProfileManager` does not support them, but if you pass multiple profiles (profile names separated by a comma), everything handled by SmallRye Config will work. SmallRye Config fully supports multiple Profiles. 

In most cases, the Quarkus profile will be the same returned by SmallRye Config (except that it returns in a List type, so the first element contains the profile). Only if you try to use multiple profiles do they return different values. For instance:

```properties
quarkus.profile=common,dev
```

Quarkus will use `common,dev` as the active profile. SmallRye Config will use `common` and `dev` as active profiles.

Here is some additional information:
https://smallrye.io/smallrye-config/2.11.0/config/profiles/

**Reply by jehrenzweig-leagueapps** [reply]
Great, thank you for providing all that additional info.  I'll revisit the code I wrote originally & make sure I know how to use the `SmallRyeConfig` approach; if I have any more questions or run into any more problems, I'll update this ticket.
